src/openapi/streaming/subscription.ts

```diff
@@ -0,0 +1,1160 @@
+                    currentState: this.currentState,
+                    url: this.url,
+                    servicePath: this.servicePath,
+                });
+        }
+
+        try {
+            this.processUpdate(message, this.UPDATE_TYPE_DELTA);
+        } catch (error) {
+            log.error(
+                LOG_AREA,
+                'Exception occurred in streaming delta update callback',
+                {
+                    error: {
+                        message: error.message,
+                        stack: error.stack,
+                    },
+                    payload: message,
+                    url: this.url,
+                    servicePath: this.servicePath,
+                },
+            );
+        }
+    }
+
+    /**
+     * Handles a heartbeat from the server.
+     */
+    onHeartbeat() {
+        if (this.currentState === this.STATE_SUBSCRIBE_REQUESTED) {
+            log.debug(
+                LOG_AREA,
+                'Received heartbeat for a subscription we havent subscribed to yet',
+                { url: this.url, servicePath: this.servicePath },
+            );
+        }
+        this.onActivity();
+    }
+
+    /**
+     * Handle a subscription pending unsubscribe by tag.
+     */
+    onUnsubscribeByTagPending() {
+        this.tryPerformAction(ACTION_UNSUBSCRIBE_BY_TAG_PENDING);
+    }
+
+    /**
+     * Handled a subscription having been unsubscribed by tag.
+     */
+    onUnsubscribeByTagComplete() {
+        this.setState(this.STATE_UNSUBSCRIBED);
+        this.onReadyToPerformNextAction();
+    }
+
+    /**
+     * Returns whether this subscription is ready to be unsubscribed by tag after it has been requested.
+     */
+    isReadyForUnsubscribeByTag() {
+        return this.currentState === this.STATE_READY_FOR_UNSUBSCRIBE_BY_TAG;
+    }
+
+    /**
+     * Returns the time in ms till the subscription would be orphaned.
+     * @param now - The current time as a reference (e.g. Date.now()).
+     */
+    timeTillOrphaned(now: number) {
+        // this works because there are no suspended and resume states.
+        // once subscribed, orphan finder will be notified.
+        if (
+            !this.connectionAvailable ||
+            this.inactivityTimeout === 0 ||
+            this.currentState === this.STATE_UNSUBSCRIBED ||
+            this.currentState === this.STATE_UNSUBSCRIBE_REQUESTED ||
+            this.currentState === this.STATE_SUBSCRIBE_REQUESTED
+        ) {
+            return Infinity;
+        }
+
+        // Follows the same pattern as the old library, not giving any grace period for receiving a heartbeat
+        // if it was required, it could be added on here
+
+        // @ts-expect-error FIXME this.latestActivity  may be undefined since it was not
+        // initialized in the old code - verify whether it can be undefined at this point or initialized with 0 in  the constructor
+        const diff = now - this.latestActivity;

```

Comment:
not sure it can happen but seems safe to initialise it. maybe to Date.now? that makes the diff small which makes the return value larger which is correct - at start time till orphaned should be big

Reply:
this method is used in `orphaFinder`, and with the current implementation, in case `latestActivity` or `inactivityTimeout` is `undefined`, the `NaN` is returned:
```ts
const timeTillOrphaned = subscription.timeTillOrphaned(now);
if (timeTillOrphaned <= 0) {
    orphanedSubscriptions.push(subscription);
} else if (timeTillOrphaned < newNextUpdateIn) {
    foundNextUpdate = true;
    newNextUpdateIn = timeTillOrphaned;
}
```

 so if we return `Infinity` the same behavior should be preserved. I made it like this in the latest commit. Let me know if there might be an issue with this.